File: Core kernel razonamiento repo para Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/dotnet/src/GitHubTokenProvider.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

using System.Diagnostics.CodeAnalysis;

namespace GitHub.Copilot;

/// <summary>Why the runtime is requesting a GitHub token.</summary>
[Experimental(Diagnostics.Experimental)]
public enum GitHubTokenRequestReason
{
    /// <summary>The session needs its initial token.</summary>
    Initial,

    /// <summary>The session needs a refreshed token.</summary>
    Refresh,
}

/// <summary>Arguments passed to a session-scoped GitHub token provider.</summary>
[Experimental(Diagnostics.Experimental)]
public sealed class GitHubTokenProviderArgs
{
    /// <summary>Gets the effective GitHub host for which a token is needed.</summary>
    public required string Host { get; init; }

    /// <summary>
    /// Gets the session receiving the token, or <see langword="null"/> before a
    /// cloud session has been assigned an identifier.
    /// </summary>
    public string? SessionId { get; init; }

    /// <summary>Gets whether the runtime needs an initial or refreshed token.</summary>
    public required GitHubTokenRequestReason Reason { get; init; }
}

/// <summary>A GitHub access token returned by a session-scoped provider.</summary>
[Experimental(Diagnostics.Experimental)]
public sealed class GitHubToken
{
    /// <summary>Gets or sets the GitHub access token.</summary>
    public required string AccessToken { get; set; }

    /// <summary>Gets or sets the OAuth token type. The runtime defaults it to <c>bearer</c>.</summary>
    public string? TokenType { get; set; }

    /// <summary>
    /// Gets or sets the required positive number of seconds remaining when the
    /// callback completes. Production GitHub tokens typically last eight hours.
    /// </summary>
    public required long ExpiresIn { get; set; }

    /// <inheritdoc />
    public override string ToString()
        => $"{nameof(GitHubToken)} {{ {nameof(TokenType)} = {TokenType}, {nameof(ExpiresIn)} = {ExpiresIn}, {nameof(AccessToken)} = <redacted> }}";
}

/// <summary>The result returned by a session-scoped GitHub token provider.</summary>
[Experimental(Diagnostics.Experimental)]
public sealed class GitHubTokenProviderResult
{
    /// <summary>Gets whether token acquisition was cancelled.</summary>
    public bool Cancelled { get; private init; }

    /// <summary>Gets the acquired token, if acquisition was not cancelled.</summary>
    public GitHubToken? Token { get; private init; }

    /// <summary>Creates a successful token result.</summary>
    public static GitHubTokenProviderResult FromToken(GitHubToken token)
    {
        ArgumentNullException.ThrowIfNull(token);
        return new() { Token = token };
    }

    /// <summary>Creates a cancelled result.</summary>
    public static GitHubTokenProviderResult Cancel() => new() { Cancelled = true };
}

File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/go/github_token_provider.go
package copilot

import (
	"fmt"

	"github.com/github/copilot-sdk/go/rpc"
)

// GitHubTokenRequestReason describes why the runtime needs a GitHub token.
//
// Experimental: GitHubTokenRequestReason may change or be removed.
type GitHubTokenRequestReason = rpc.GitHubTokenAcquireReason

const (
	// GitHubTokenRequestReasonInitial indicates the session needs its initial token.
	GitHubTokenRequestReasonInitial = rpc.GitHubTokenAcquireReasonInitial
	// GitHubTokenRequestReasonRefresh indicates the session needs a refreshed token.
	GitHubTokenRequestReasonRefresh = rpc.GitHubTokenAcquireReasonRefresh
)

// GitHubTokenProviderArgs contains the context for a GitHub token request.
//
// Experimental: GitHubTokenProviderArgs may change or be removed.
type GitHubTokenProviderArgs struct {
	// Host is the effective GitHub host for which a token is needed.
	Host string
	// SessionID identifies the session receiving the token. It is nil before a
	// cloud session has been assigned an ID.
	SessionID *string
	// Reason indicates whether this is the initial token or a refresh.
	Reason GitHubTokenRequestReason
}

// GitHubToken contains a GitHub access token returned by a provider.
//
// Experimental: GitHubToken may change or be removed.
type GitHubToken struct {
	// AccessToken is the GitHub access token.
	AccessToken string
	// TokenType is the OAuth token type. The runtime defaults it to "bearer".
	TokenType *string
	// ExpiresIn is the required positive number of seconds remaining when the
	// callback completes. Production GitHub tokens typically last eight hours.
	ExpiresIn int64
}

// String returns a redacted description that never includes the access token.
func (t GitHubToken) String() string {
	tokenType := ""
	if t.TokenType != nil {
		tokenType = *t.TokenType
	}
	return fmt.Sprintf("GitHubToken{TokenType:%q, ExpiresIn:%d, AccessToken:<redacted>}", tokenType, t.ExpiresIn)
}

// GoString returns a redacted Go-syntax description that never includes the access token.
func (t GitHubToken) GoString() string {
	return t.String()
}

// GitHubTokenProviderResult is the result of a GitHub token request.
//
// Experimental: GitHubTokenProviderResult may change or be removed.
type GitHubTokenProviderResult struct {
	Cancelled bool
	Token     *GitHubToken
}

// GitHubTokenResult returns a successful token-provider result.
func GitHubTokenResult(token *GitHubToken) *GitHubTokenProviderResult {
	return &GitHubTokenProviderResult{Token: token}
}

// GitHubTokenCancelled returns a result indicating that token acquisition was cancelled.
func GitHubTokenCancelled() *GitHubTokenProviderResult {
	return &GitHubTokenProviderResult{Cancelled: true}
}

// GitHubTokenProvider acquires session-scoped GitHub tokens on demand. Initial
// cancellation, errors, and invalid token responses reject session creation or
// resume instead of falling back to ambient authentication.
//
// Experimental: GitHubTokenProvider may change or be removed.
type GitHubTokenProvider func(args GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error)
